text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+an error during this time. This is meant to prevent infinite rerenders and other
+antipatterns.
+
+This hook is only called for helpers with the `isScheduledEffect` capability
+enabled. It has the following timing semantics:
+
+**Always**
+- called after the helper was first created
+- called after autotracked state has changed
+
+**Never**
+- called if the `isScheduledEffect` capability is disabled
+
+#### `getDestroyable`
+
+`getDestroyable` is an optional hook that users can use to register a
+destroyable object for the helper. This destroyable will be registered to the
+containing block or template parent, and will be destroyed when it is destroyed.
+See the [Destroyables RFC](https://github.com/emberjs/rfcs/blob/master/text/0580-destroyables.md)
+for more details.
+
+`getDestroyable` is only called if the `hasDestroyable` capability is enabled.
+
+This hook has the following timing semantics:
+
+**Always**
+- called immediately after the `createHelper` hook is called
+
+**Never**
+- called if the `hasDestroyable` capability is disabled
+
+### Capabilities
+
+There are three proposed capabilities for helper managers:
+
+* `hasDestroyable`
+* `isScheduledEffect`
+* `hasValue`
+
+Out of these capabilities, one of `isScheduledEffect` or `hasValue` _must_ be
+enabled. The other must _not_ be enabled, meaning they are mutually exclusive.
+
+#### `hasDestroyable`
+
+Determines if the helper has a destroyable to include in the destructor
+hierarchy. If enabled, the `getDestroyable` hook will be called, and its result
+will be associated with the destroyable parent block.
+
+#### `hasValue`
+
+Determines if the helper has a value which can be used externally. The helper's
+`getValue` hook will be run whenever the value of the helper is accessed if this
+capability is enabled.
+
+#### `isScheduledEffect`
+
+Determines if the helper is scheduled. If enabled, the helper's `runEffect` hook
+will run after render, and will not allow any type of state mutation when
+running.
+
+### Scheduled Helpers Timing
+
+Scheduled helpers run their effects after render, and after modifiers have been
+applied for a given render, but before paint. The exact timing may shift around,
+and may or may not correspond to a single rendering pass in cases where there
+are multiple rendering passes in a single paint.
+
+In the future different timings may be added as options for scheduling. For
+instance, a timing to call the effect using `requestIdleCallback`, when the
+browser has finished rendering and handling higher priority work, could be
+added. However, this is out of scope for this RFC.
+
+## How we teach this
+
+Helper managers are a low-level construct that is generally only meant to be
+used by experts and addon authors. As such, it will only be taught through API
+documentation. In addition, for precision and clarity, the API docs will include
+snippets of TypeScript interfaces where appropriate.
```

Comment:
This is good, but I think probably somewhat incomplete. These should certainly be covered in <abbr title="application programming interface">API</abbr> docs, but <abbr title="application programming interface">API</abbr> docs are insufficient for actually building the appropriate mental model for consumers. (Both the CLI and Data docs here are prime examples of this problem historically for Ember! Even when they had full coverage of the API surface, it has been difficult to *learn* or *master*.)

[Effective documentation](https://documentation.divio.com/introduction/) includes not only <i>reference materials</i> but also <i>tutorials</i> (introducing the , <i>how-to-guides</i> (like “cookbooks”), and <i>explanations</i>. This RFC itself would suffice for the *explanation*, and <abbr title="application programming interface">API</abbr> docs will handle <i>reference materials</i>, but I think we should also identify what appropriate kinds of <i>tutorial</i> and <i>how-to</i> material are appropriate here—presumably in an <b>Advanced Guides</b> section for addon authors specifically.

Reply:
The issue here is this is not even a "standard" advanced Ember topic. I would say the current advanced section (In-Depth Topics) is currently stuff that most intermediate and senior Ember devs will eventually need to know or learn. Ideally, _most_ Ember apps will never include a custom manager directly, and most senior Ember devs will not _have_ to write one. 

I think a guide would be useful, and I'd be happy to write one for managers in general, but I'm not sure where that would live. Maybe "Advanced Topics" would be good to add to the guides, we've shied away from the term because we didn't want to discourage beginners who are curious (thus the name "in-depth") but this would be a time when that would make sense.

Reply:
Just wanted to pitch in that this is something the learning team has considered previously and recently discussed at one of the meeting. If there's interest enough interest to kick-off an RFC…

Reply:
I’d be happy to co-author, @locks! I actually mentioned as much to @pzuraq in a DM very recently!